coderd/parameters.go

```diff
@@ -71,48 +99,40 @@ func (api *API) templateVersionDynamicParameters(rw http.ResponseWriter, r *http
 		return
 	}
 
+	// Add the file first. Calling `Release` if it fails is a no-op, so this is safe.
 	templateFS, err := api.FileCache.Acquire(fileCtx, fileID)
+	defer api.FileCache.Release(fileID)
 	if err != nil {
 		httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
 			Message: "Internal error fetching template version Terraform.",
 			Detail:  err.Error(),
 		})
 		return
 	}
-	defer api.FileCache.Release(fileID)
 
 	// Having the Terraform plan available for the evaluation engine is helpful
 	// for populating values from data blocks, but isn't strictly required. If
 	// we don't have a cached plan available, we just use an empty one instead.
 	plan := json.RawMessage("{}")
-	tf, err := api.Database.GetTemplateVersionTerraformValues(ctx, templateVersion.ID)
-	if err == nil {
+	if len(tf.CachedPlan) > 0 {
 		plan = tf.CachedPlan
+	}
 
-		if tf.CachedModuleFiles.Valid {
-			moduleFilesFS, err := api.FileCache.Acquire(fileCtx, tf.CachedModuleFiles.UUID)
-			if err != nil {
-				httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
-					Message: "Internal error fetching Terraform modules.",
-					Detail:  err.Error(),
-				})
-				return
-			}
-			defer api.FileCache.Release(tf.CachedModuleFiles.UUID)
-			templateFS = files.NewOverlayFS(templateFS, []files.Overlay{{Path: ".terraform/modules", FS: moduleFilesFS}})
+	if tf.CachedModuleFiles.Valid {
+		moduleFilesFS, err := api.FileCache.Acquire(fileCtx, tf.CachedModuleFiles.UUID)
+		defer api.FileCache.Release(tf.CachedModuleFiles.UUID)
```

Comment:
this is a double free if `err != nil`

Reply:
Ah I saw you made a change so error != nil calls Release :+1: